Add cross domain message test

NingLin-P opened the pull request:
This PR adds a cross domain message test, besides that this PR also:
- ~~Fixed an issue introduced in https://github.com/subspace/subspace/commit/146b9078c2e2bee225632604375143278f09df07, which will cause the relayer to miss some blocks.~~
- Introduce some helper macro to make the test cleaner.


### Code contributor checklist:
* [x] I have read, understood and followed [contributing guide](https://github.com/subspace/subspace/blob/main/CONTRIBUTING.md)

NingLin-P committed:
Fix typo in domain_test_service::core_domain

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Add cross domain message support for system-domain-test-runtime

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Introduce OnchainStateApi for domain testing

OnchainStateApi contains api that used to inspect onchain state which is helpful in
test to assert some states are upated according to the test.

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Introduce produce_blocks macro to replace futures::join in test

When using futures::join if the primary node fail to produce block the whole future will stuck
because the domain node will also not able to produce block, which hurt debug efficiency a lot.

This commit also come with the produce_blocks_until macro which will be used in the upcome test.

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Add test test_cross_domains_message_should_work

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Set up mock primary node logging properly

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Decrease RelayConfirmationDepth in test to accelerate test

Signed-off-by: linning <linningde25@gmail.com>

vedhavyas commented on crates/subspace-fraud-proof/src/tests.rs:
```diff
@@ -127,13 +127,7 @@ async fn execution_proof_creation_and_verification_should_work() {
-    )
-    .2
-    .unwrap();
+    produce_blocks!(ferdie, alice, bob, 1).await.unwrap();
```
I have mixed feelings about this. We had a clear distinction between, who is producing and who is waiting. Not feels like everyone is producing blocks. Not a Blocker though

NingLin-P replied:
The domain nodes are also producing blocks, but they are subject to the primary node instead of producing blocks proactively, thus we are using `wait_for_blocks` internally.

vedhavyas commented on domains/client/domain-executor/src/tests.rs:
```diff
@@ -1095,3 +1051,95 @@ async fn existing_bundle_can_be_resubmitted_to_new_fork() {
+                dst_domain_id: DomainId::SYSTEM,
+                params: InitiateChannelParams {
+                    max_outgoing_messages: 100,
+                    fee_model: Default::default(),
```
I usually put some number to the fees instead of zero so we may also find other issues regarding the balances

vedhavyas commented on domains/client/domain-executor/src/tests.rs:
```diff
@@ -1095,3 +1051,95 @@ async fn existing_bundle_can_be_resubmitted_to_new_fork() {
+        let post_alice_free_balance = alice.free_balance(alice.key.to_account_id());
+        let post_bob_free_balance = bob.free_balance(bob.key.to_account_id());
+
+        post_alice_free_balance == pre_alice_free_balance - transfer_amount
```
this needs to be adjusted if we introduce fees

vedhavyas reviewed (commented):
Looks good overall. Seem like we only test XDM between System domain <> Core domain. We also need to test between core domains. Do we not have the infra to spin up multiple core domains at a time ?

NingLin-P committed:
Add core eth relay domain test runtime for testing

It is mostly copied and pasted from the production runtime with a few minor changes:
- RelayConfirmationDepth is decreased to 1 to accelerate test
- Impl the OnchainStateApi which is required by the upcoming test

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Add support to build core eth relay node in domain-test-service

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Add test case for cross domain msg between core domains

Signed-off-by: linning <linningde25@gmail.com>

vedhavyas reviewed (approved):
would have been nice if Core EVM was used. Not a blocker. LGTM

NingLin-P committed:
Merge branch 'main' into add-xdm-test
